pipeline/astro_cal.py

```diff
@@ -28,11 +149,472 @@ class AstroCalibrator:
+        if minmag is not None:
+            gaia_query += f"AND phot_g_mean_mag>={minmag} "
+        if maxmag is not None:
+            gaia_query += f"AND phot_g_mean_mag<={maxmag} "
+        _logger.debug( f'gaia_query is "{gaia_query}"' )
+
+        countdown = 5
+        while countdown > 0:
+            try:
+                qresult = queryClient.query( sql=gaia_query )
+                countdown = 0
+            except Exception as e:
+                countdown -= 1
+                if countdown > 0:
+                    _logger.info( f"Sleeping 5s and retrying gaia query after failed attempt ({str(e)})" )
+                    time.sleep(5)
+                    continue
+                else:
+                    _logger.info( f"Gaia query failed attempt ({str(e)}" )
+                    err = "Giving up after repeated failures querying NOIRLab for gaia stars"
+                    raise RuntimeError( err )
+        df = dl.helpers.utils.convert( qresult, "pandas" )
+
+        # Convert this into a FITS file format that scamp would recognize.
+        # In particular, FITS header keywords have to be all upper case
+        # The _WORLD variables are standard.
+        # The others are not global standards, so we'll have to do things
+        # consistent with what we'll do in scamp.
+        # NOTE : the mag error columns are wrong by a factor of 1.09 = 2.5/ln(10)
+
+        df.rename(
+            columns={
+                "ra": "X_WORLD",
+                "dec": "Y_WORLD",
+                "ra_error": "ERRA_WORLD",
+                "dec_error": "ERRB_WORLD",
+                "phot_g_mean_mag": "MAG_G",
+                "phot_g_mean_flux_over_error": "MAGERR_G",
+                "phot_bp_mean_mag": "MAG_BP",
+                "phot_bp_mean_flux_over_error": "MAGERR_BP",
+                "phot_rp_mean_mag": "MAG_RP",
+                "phot_rp_mean_flux_over_error": "MAGERR_RP",
+                "pm": "PM",
+                "pmra": "PMRA",
+                "pmdec": "PMDEC" },
+            inplace=True
+        )
+        # Make the errors actual magnitude errors.  (1.0857 = 2.5/ln(10))
+        for band in [ 'G', 'BP', 'RP' ]:
+            df[ f'MAGERR_{band}' ] = 1.0857 / df[ f'MAGERR_{band}' ]
+        # Put in two more fields that scamp really wants
+        df[ 'OBSDATE' ] = 2015.5
+        df[ 'FLAGS' ] = 0
+
+        # To avoid saving too many files in one directory, use a healpix subdirectory
+        hpix = healpy.ang2pix( 4, ra, dec, lonlat=True )
+        minmagstr = 'None' if minmag is None else f'{minmag:.1f}'
+        maxmagstr = 'None' if maxmag is None else f'{maxmag:.1f}'
+        relpath = ( pathlib.Path( "GaiaDR3_excerpt" ) / str(hpix) /
+                    f"Gaia_DR3_{ra:.4f}_{dec:.4f}_{minmagstr}_{maxmagstr}.fits" )
+        ofpath = pathlib.Path( FileOnDiskMixin.temp_path ) / relpath
+        dbpath = pathlib.Path( FileOnDiskMixin.local_path ) / relpath
+        ofpath.parent.mkdir( parents=True, exist_ok=True )
```

Comment:
I'm not sure what the advantage would be, since we already know that it's going underneath temp_dir, which is a legal place, so there's no need to verify it's going to a legal place.

Reply:
It probably makes no difference, but if you get into the habit of always using safe_mkdir you will never find yourself writing into the home directory or whatever. Also saves some typing when you need to do `parents=True` and `exist_ok=True` which are on by default.